bpo-38294: Add list of no-longer-escaped chars to re.escape documentation

rbanffy opened the pull request:
The changes in behavior from 3.6 to 3.7 caused some minor issues with code that still needs to run on 2.7 while it is being migrated to 3.7. 

Since "/" is usually a common delimiter for regexes, we assumed it'd be escaped, as it was up to 3.6, but since that's no longer the case, it'd be nice to emphasise the difference a little bit more than the current documentation does. As it is, it's easy to miss this specific character.

<!-- issue-number: [bpo-38294](https://bugs.python.org/issue38294) -->
https://bugs.python.org/issue38294
<!-- /issue-number -->

Commit:
bpo-38294: Explicitly list characters re.escape no longer escapes

Prior to 3.7, re.escape escaped many characters that don't have
special meaning in Python, but that use to require escaping in other
tools and languages. This commit aims to make it clear which characters
were, but are no longer escaped.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Commit:
bpo-38294: Make the second example list all punctuation that is - and isn't - escaped

The current example shows escaping only of infix math operators. This example uses
string.punctuation as input and doesn't change the order of the characters for the
output.

serhiy-storchaka commented on Doc/library/re.rst:
```diff
@@ -938,9 +938,8 @@ form.
-      >>> operators = ['+', '-', '*', '/', '**']
-      >>> print('|'.join(map(re.escape, sorted(operators, reverse=True))))
-      /|\-|\+|\*\*|\*
+      >>> print(' '.join(map(re.escape, string.punctuation)))
```
I prefer to keep the existing example. It is closer to real world code and contains some non-trivial tricks which you can copy to your code even if you don't understand it.

All these examples are simplified extractions from the real code and demonstrate different cases of using `re.escape()`.

rbanffy replied:
I appreciate that, but the reverse sort distracts from what's being done to the operators list, and concatenating with pipes makes it harder to read, which makes it easier to miss. Since we don't expect alphabetical characters to be escaped, it could alternatively replace the first code example (where the characters are concatenated without spaces between them, which makes it harder to read)

We could probably go with:

```python

print(' '.join([re.escape(c) for c in string.punctuation])

```

if we judge list comprehensions may be more readable in this case. We already have a map example in the lot.

Commit:
bpo-38294: Change the string being escaped in the first example

This example illustrates two characters that changed behavior recently in
addition to ".", which continues to be escaped.

rbanffy commented on Doc/library/re.rst:
```diff
@@ -931,8 +931,8 @@ form.
 
-      >>> print(re.escape('python.exe'))
-      python\.exe
+      >>> print(re.escape('http://www.python.org'))
```
With this change, the example illustrates some of the characters that are no longer being escaped in a somewhat important use case. I think that, with it, the total change can be minimal and the other examples could be kept as they were. Your input is appreciated, @serhiy-storchaka.

serhiy-storchaka replied:
I like this. The example demonstrates the use for an arbitrary literal string. It contains letters, which should not be escaped, special characters which must be escaped, and characters which was escaped in old Python versions, but which are not escaped now.

Please restore the third example.

Commit:
bpo-38294: Restore original version of the third example.

serhiy-storchaka commented on Doc/tools/susp-ignored.csv:
```diff
@@ -356,3 +356,4 @@ whatsnew/changelog,,::,default::BytesWarning
 whatsnew/changelog,,::,default::DeprecationWarning
 library/importlib.metadata,,:main,"EntryPoint(name='wheel', value='wheel.cli:main', group='console_scripts')"
 library/importlib.metadata,,`,loading the metadata for packages for the indicated ``context``.
+library/re,,`,"`"
```
Is this still needed?

serhiy-storchaka replied:
Just tested. It is needed.
